TST/BUG: fromfile - fix test and expose bug with io class argument

mcmtroffaes opened the pull request:
This fixes an obvious typo in a test, however in doing so it also exposes a bug on Python 2.7. The bug was originally reported at https://github.com/mcmtroffaes/pathlib2/pull/7 and I stumbled on this whilst trying to fix it. Sadly the error message is rather unhelpful:

```
SystemError: error return without exception set
```

without further traceback into the numpy source, so I haven't been able to identify the source of this error yet.

mcmtroffaes committed:
BUG: npy_PyFile_Dup2 - fix PyFile_AsFile failing on io style classes

This bug only manifests itself in the Python 2 code path.

Falls back to the current Python 3 code path also on Python 2 because
the Python 3 code path is written precisely to handle this situation.

Also fix tests, and clarify in the documentation that the current
implementation requires the stream to be seekable.

mcmtroffaes commented:
From the travis run it looks like this also exposes a bug in the Python 3 code path, I'll investigate it further when I have time.

mcmtroffaes commented:
@cjermain As you originally wrote the code and the test for allowing more general objects to be passed to fromfile (see https://github.com/numpy/numpy/pull/6330), do you see why your test fails? I'm trying to debug this with gdb, but not getting anything useful out of it yet, so I'd appreciate your expertise on this.

cjermain commented:
Thanks @mcmtroffaes for catching this -- looks like my test was written in haste. The bug fix that this tests for, #6330, solved a regression in Python 3 that did not allow unbuffered IO when their tell or seek methods were not present. This only affects the Python 3 branch (code was changed in `npy_3kcompat.h`), so its likely that the test is the issue that affects Python 2. With the test, I was trying to simulate when the tell or seek methods were not there (or raised an error on being called), as this `fromfile` should not require tell or seek to read for unbuffered IO in either Python 2 or 3. I'm not sure off the cuff why its failing, but I'll test it when I have a chance.

mcmtroffaes commented:
Thanks for the quick response @cjermain and indeed, your patch does not affect the Python 2 code path and therefore it's definitely an older bug that your test is simply teasing out. The current implementation of fromfile does require tell() however, even if the stream is unbuffered. It calls get_remaining_size:

https://github.com/numpy/numpy/blob/7036842/numpy/core/records.py#L755

which must use tell() to get the current position in the stream:

https://github.com/numpy/numpy/blob/7036842/numpy/core/records.py#L711

So fromfile, in its current implementation, only works for random access streams (but both buffered and unbuffered should work all the same, as long as the stream random access).

A better implementation of fromfile would first read the bytes from the file (without the get_remaining_size), and only then allocate the array and fill it up. We still need random access if offset > 0:

https://github.com/numpy/numpy/blob/7036842/numpy/core/records.py#L753

but that's acceptable. Anyway, I assume that would fix the Python 3 problem. I'm still figuring out what the Python 2 issue is caused by...

mcmtroffaes commented:
Looking more closely at the implementation, we need to know the number of remaining bytes before the data is read because the data is read directly into the array, and the array size obviously depends on the number of bytes remaining. I've updated the documentation to mention that random access is really required, and "fixed" the test so it no longer fails on tell and seek. This should fix the Python 3 tests on travis.

The only way to avoid random access would be to only allow fromfile to read entire files (i.e. it would be the caller's responsibility to ensure that the file is at position zero), because only in that case we can know the number of remaining bytes without using tell or seek before reading in the data (as we can get the size via os.stat). That would be a behavioural change and break backwards compatibility however.

mcmtroffaes commented:
Just to confirm that the array fromfile has a similar limitation:

https://github.com/numpy/numpy/blob/7036842/numpy/core/src/multiarray/multiarraymodule.c#L2083
https://github.com/numpy/numpy/blob/7036842/numpy/core/src/multiarray/ctors.c#L3366
https://github.com/numpy/numpy/blob/7036842f828fa41f95c333812bccccb3ae469f91/numpy/core/src/multiarray/ctors.c#L3217

So the Python 3 implementation is still using seek & tell if you don't tell it how many elements to get because in that case num equals -1.

mcmtroffaes commented:
Ok, bug tracked & fixed, tests are passing finally. @cjermain does this look good to you?

cjermain commented:
Yes, it looks good. I did a manual test to make sure that wking/pycomedi#3 Python 3 compatibility works. Thanks @mcmtroffaes!

mcmtroffaes commented:
Many thanks for looking over it and doing the extra testing @cjermain. Is there anything else needed before this can be merged?

charris commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -314,26 +314,6 @@ npy_PyFile_Check(PyObject *file)
 
-#else
-
-static NPY_INLINE FILE *
```
Why the removal here?

mcmtroffaes replied:
The Python 2 code path is broken because ``PyFile_AsFile(file)`` doesn't work when ``file`` is an ``IOFile`` (as produced by ``io.open``). So the patch removes this code path and simply uses the old Python 3 code path also on Python 2, because that's exactly what the Python 3 code path does: creating ``FILE *`` objects from things like ``IOFile`` and so on.

charris commented:
@mcmtroffaes Could you squash the 8 commits into one?

mcmtroffaes commented:
I've rebased it on the current master (it no longer merged without conflicts) and squashed the commits. The commit message says "BUG" but this also affects "DOC" and "TST" - I assume this is ok but if you want it differently just give a shout.

mcmtroffaes commented:
The appveyor test is now failing (oddly it was passing before); this will need some more investigation on windows.

pv commented:
The Python 3 code path itself does not necessarily cover cases that PyFile_AsFile does, eg. writing to pipes and so on. PyFile_AsFile is the best option available for old-style file objects --- the io file classes could perhaps be detected and handled separately.

mcmtroffaes commented:
Ok, I'll see if I can resurrect PyFile_AsFile for old-style file objects for Python 2.

homu commented:
:umbrella: The latest upstream changes (presumably #7201) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
@mcmtroffaes Any progress?

mcmtroffaes commented:
Sorry for the silence on this one - I've rebased to resolve the conflict. Still need to address https://github.com/numpy/numpy/pull/6632#issuecomment-171265394

mcmtroffaes commented:
... and some tests on windows are failing, will look at these.

homu commented:
:umbrella: The latest upstream changes (presumably #7099) made this pull request unmergeable. Please resolve the merge conflicts.

mcmtroffaes commented:
I've rebased once more. Still some windows failures on Python 2.7 (oddly these tests pass on linux...). Will look into these.

seberg commented:
Seems a bit like the finally clause has to make sure the file was created if it wants to close it. After fixing that, the error message should at least be a bit more useful....

charris commented:
I'm going to move this to 1.13.0. If you get things fixed make a comment and I will try to get it into 1.12.0 if the timing is good.

mcmtroffaes commented:
Windows tests are passing now, and I've addressed https://github.com/numpy/numpy/pull/6632#issuecomment-171265394 so patch is now pretty much ready.

There's one last failure (a new one) on travis which I can't figure out how to fix. The compiler warns that orig_pos, as used here:

https://github.com/numpy/numpy/pull/6632/files#diff-af042e24f2e36d598c1b149fbb5df756R288

may be uninitialized. However orig_pos is a parameter of the function, and the patch has nothing to do with it as far as I can tell... Direct link to travis:

https://travis-ci.org/numpy/numpy/jobs/158421508

Any ideas?

mcmtroffaes commented:
Oh I figured it out - orig_pos is shadowing another declaration from the file it's included from. Renaming should fix it...

mcmtroffaes committed:
BUG: fix uninitialized variable

mcmtroffaes commented:
Tests are green! Should be good now for final review.

charris commented:
OK, I think this is OK. @pv, @juliantaylor Comment?

juliantaylor commented:
looks good

juliantaylor commented:
needs rebasing to get our updated tests

charris commented:
Yep, rebasing fixes the failing test, see #8947. Given the change set, I think it should be safe to merge. Hmm, I going to try and experimental close and reopen to see what happens.

charris commented:
That fixed it. Apparently rerunning the test isn't enough to get a new merge.

@mcmtroffaes Merged, thanks,
